### PR TITLE
[FLOC-1724] Sketch of logging options.

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-1724

### CODE REVIEW

**Comment by adamtheturtle** [comment]
Some build failures (e.g. `lint`) are relevant.

File: flocker/common/script.py [hunk 1]
```diff
@@ -26,6 +29,23 @@
...
+    :cvar DISCARD: Discard all logs.
+    :cvar STDOUT: Log to standard out.
+    :cvar FILE: Log to a file. Defaults to a file in XDG_CACHE_HOME, but
+        overrideable on the command line.
+    """
+    # Used by legacy flocker-volume
+    DISCARD = NamedConstant()
+    # Used by daemons (flocker-control, flocker-*-agent)
+    STDOUT = NamedConstant()
+    # Used by CLI scripts (flocker, flocker-ca, flocker-deploy, flocker-provision)
```
Comment (adamtheturtle) [inline flocker/common/script.py]: Put "e.g." else this will soon be out of date.

File: flocker/common/script.py [hunk 2]
```diff
@@ -160,7 +191,8 @@ def main(self):
         # always do this first before any side-effecty code is run:
         options = self._parse_options(self.sys_module.argv[1:])
 
-        if self.logging:
+        # TODO Factor this out to dispatch on LoggingPolicy
```
Comment (adamtheturtle) [inline flocker/common/script.py]: Do this, or create a follow-up.

Comment (adamtheturtle) [inline flocker/common/script.py]: Actually do it here, it isn't a big task.

File: flocker/common/script.py [hunk 3]
```diff
@@ -26,6 +29,23 @@
 ]
 
 
+class LoggingPolicy(Names):
```
Comment (adamtheturtle) [inline flocker/common/script.py]: This is missing tests.

File: flocker/common/script.py [hunk 4]
```diff
@@ -132,6 +159,10 @@ def __init__(self, script, options, logging=True,
             reactor = global_reactor
         self._reactor = reactor
 
+        if logging is LoggingPolicy.FILE:
+            # This adds log-dir (and verbose?) to the provided options.
```
Comment (adamtheturtle) [inline flocker/common/script.py]: Clarify what this does (no "?").

File: flocker/common/script.py [hunk 5]
```diff
@@ -170,6 +202,12 @@ def main(self):
             # contrast needs to be shutdown because it starts a thread
             # that will keep the process from exiting.
             observer.start()
+        elif self.logging is LoggingPolicy.STDOUT:
+            log_dir = FilePath(self.options['log-dir'])
+            log_file = log_dir.child(
```
Comment (adamtheturtle) [inline flocker/common/script.py]: Write a comment or refactor this so that what it is doing is clear.

File: flocker/common/script.py [hunk 6]
```diff
@@ -53,12 +73,19 @@ def opt_version(self):
...
+    # that is only applied to CLI scripts
     def opt_verbose(self):
         """Turn on verbose logging."""
         self['verbosity'] += 1
     cls.opt_verbose = opt_verbose
     cls.opt_v = opt_verbose
 
+    cls.optParameters = getattr(cls, 'optParameters', [])
+    cls.optParameters.append(
+        ['log-dir', None, user_log_dir("Flocker", "ClusterHQ"),
```
Comment (adamtheturtle) [inline flocker/common/script.py]: I'm not sure what a good default log location is so I'd like someone else to look at this.

**Comment by adamtheturtle** [comment]
Ah shucks. I started doing a code review before seeing that I should be doing a design review.

**Comment by adamtheturtle** [comment]
I'm going to put this back as the comments I had which were like "someone else should look at this" are the relevant ones for design review.

File: flocker/common/script.py (hunk 5 above)
Comment (itamarst) [inline flocker/common/script.py]: Looks like this was supposed to be `LoggingPolicy.FILE`?

**Comment by itamarst** [comment]
So if I understand this correctly:

1. Each script has a policy: no logging, stdout, or file.
2. File policy logs to a default location which can be overrriden.
3. It's not clear what verbose does.

I propose a slight tweak which I think improves usability. There are three modes:

1. Stdout. No verbose flag.
2. No logging by default. There is a verbose or enable-logging flag which enables logging to a file, and therefore also command-line option to determine where logs go.
3. Logging to file. There is a command-line option to determine where logs go. No verbose flag.

**Comment by itamarst** [comment]
On further thought, since command-line tools aren't long running and won't product much logs, I suggest not bothering with second policy and just having all scripts either log stdout or to file, and drop the verbose flag.

**Comment by tomprince** [comment]
The only reason for having a "no logging" option is for the `flocker-volumes` script which is called remotely by the running system. I didn't think the XDG_USER_DIR of root was a good place for the logs, so wanted to separate it out. Since it is going away, I figured I'd just drop the logs for the time being.

**Comment by tomprince** [comment]
Regarding getting rid of verbose, I agree whole-heartedly.

**Comment by itamarst** [comment]
Re `flocker-volume` do whatever takes the least amount of work. If it logs something to file that's fine.

**Pull request closed without merging**